File: services/lpo-service/utils/coa_client.py
"""
Chart of Accounts Service Client

This module provides a client for interacting with the Chart of Accounts service
to create accounts and journal entries for double-entry bookkeeping.
"""

import os
import logging
from typing import List, Dict, Any, Optional
from datetime import datetime
import httpx

logger = logging.getLogger(__name__)


class CoAClient:
    """Client for Chart of Accounts service"""
    
    def __init__(self, base_url: Optional[str] = None):
        self.base_url = base_url or os.getenv("COA_SERVICE_URL", "http://chart-of-accounts-service:8080")
        self.timeout = 30.0
    
    def _get_headers(self, tenant_id: str, user_id: str, user_role: str) -> Dict[str, str]:
        """Build headers for CoA service requests"""
        return {
            "Content-Type": "application/json",
            "X-Tenant-ID": tenant_id,
            "x-keycloak-id": user_id,
            "X-User-Role": user_role,
        }
    
    async def create_account(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        code: str,
        name: str,
        account_type: str,
        currency: str = "NGN",
        parent_id: Optional[str] = None,
        description: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Create a new account in the Chart of Accounts
        
        Args:
            tenant_id: Tenant identifier
            user_id: User making the request
            user_role: User role (e.g., bank_admin, finance_admin)
            code: Account code (e.g., "1400")
            name: Account name (e.g., "Loans Receivable")
            account_type: Account type (asset, liability, equity, revenue, expense)
            currency: Currency code (default: NGN)
            parent_id: Parent account ID for hierarchical accounts
            description: Optional account description
            
        Returns:
            Created account data
        """
        account_data = {
            "code": code,
            "name": name,
            "type": account_type,
            "currency": currency,
            "is_active": True,
            "is_system_account": True,
        }
        
        if parent_id:
            account_data["parent_id"] = parent_id
        if description:
            account_data["description"] = description
        
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            try:
                response = await client.post(
                    f"{self.base_url}/api/v1/accounts",
                    json=account_data,
                    headers=self._get_headers(tenant_id, user_id, user_role),
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Failed to create CoA account: {e}")
                raise
    
    async def create_journal_entry(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        description: str,
        lines: List[Dict[str, Any]],
        reference: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Create a new journal entry
        
        Args:
            tenant_id: Tenant identifier
            user_id: User making the request
            user_role: User role (e.g., bank_admin, finance_admin)
            description: Entry description
            lines: List of journal lines with account_id, description, debit_amount, credit_amount
            reference: Optional reference (e.g., transaction ID)
            metadata: Optional metadata dictionary
            
        Returns:
            Created journal entry data
        """
        entry_data = {
            "date": datetime.utcnow().isoformat(),
            "description": description,
            "posted_by": user_id,
            "lines": lines,
        }
        
        if reference:
            entry_data["reference"] = reference
        if metadata:
            entry_data["metadata"] = metadata
        
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            try:
                response = await client.post(
                    f"{self.base_url}/api/v1/journal-entries",
                    json=entry_data,
                    headers=self._get_headers(tenant_id, user_id, user_role),
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Failed to create journal entry: {e}")
                raise
    
    async def get_accounts(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
    ) -> List[Dict[str, Any]]:
        """Get all accounts for a tenant"""
        async with httpx.AsyncClient(timeout=self.timeout) as client:
            try:
                response = await client.get(
                    f"{self.base_url}/api/v1/accounts",
                    headers=self._get_headers(tenant_id, user_id, user_role),
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Failed to get accounts: {e}")
                raise
    
    # ===== Payment Processing Specific Methods =====
    
    async def record_payment(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        payment_id: str,
        amount: int,  # Amount in kobo (smallest currency unit)
        from_account: str,
        to_account: str,
        description: str,
    ) -> Dict[str, Any]:
        """
        Record a payment transaction
        
        Debit: Receiver account
        Credit: Sender account
        """
        lines = [
            {
                "account_id": to_account,
                "description": f"Payment received - {description}",
                "debit_amount": amount,
                "credit_amount": 0,
            },
            {
                "account_id": from_account,
                "description": f"Payment sent - {description}",
                "debit_amount": 0,
                "credit_amount": amount,
            },
        ]
        
        return await self.create_journal_entry(
            tenant_id=tenant_id,
            user_id=user_id,
            user_role=user_role,
            description=f"Payment: {description}",
            lines=lines,
            reference=payment_id,
            metadata={
                "payment_id": payment_id,
                "source": "payment-processing-service",
                "event_type": "payment",
            },
        )
    
    async def record_escrow_deposit(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        escrow_id: str,
        amount: int,
        customer_account: str,
    ) -> Dict[str, Any]:
        """
        Record escrow deposit
        
        Debit: Escrow account (2300)
        Credit: Customer account
        """
        lines = [
            {
                "account_id": "2300",  # Escrow Liabilities
                "description": f"Escrow deposit for {escrow_id}",
                "debit_amount": amount,
                "credit_amount": 0,
            },
            {
                "account_id": customer_account,
                "description": f"Escrow deposit from customer",
                "debit_amount": 0,
                "credit_amount": amount,
            },
        ]
        
        return await self.create_journal_entry(
            tenant_id=tenant_id,
            user_id=user_id,
            user_role=user_role,
            description=f"Escrow deposit: {escrow_id}",
            lines=lines,
            reference=escrow_id,
            metadata={
                "escrow_id": escrow_id,
                "source": "escrow-service",
                "event_type": "deposit",
            },
        )
    
    async def record_escrow_release(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        escrow_id: str,
        amount: int,
        beneficiary_account: str,
    ) -> Dict[str, Any]:
        """
        Record escrow release
        
        Debit: Beneficiary account
        Credit: Escrow account (2300)
        """
        lines = [
            {
                "account_id": beneficiary_account,
                "description": f"Escrow release to beneficiary",
                "debit_amount": amount,
                "credit_amount": 0,
            },
            {
                "account_id": "2300",  # Escrow Liabilities
                "description": f"Escrow release for {escrow_id}",
                "debit_amount": 0,
                "credit_amount": amount,
            },
        ]
        
        return await self.create_journal_entry(
            tenant_id=tenant_id,
            user_id=user_id,
            user_role=user_role,
            description=f"Escrow release: {escrow_id}",
            lines=lines,
            reference=escrow_id,
            metadata={
                "escrow_id": escrow_id,
                "source": "escrow-service",
                "event_type": "release",
            },
        )
    
    async def record_card_transaction(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        transaction_id: str,
        amount: int,
        card_account: str,
        merchant_account: str,
        transaction_type: str,  # "purchase", "withdrawal", "refund"
    ) -> Dict[str, Any]:
        """
        Record card transaction
        
        For purchases:
          Debit: Merchant account
          Credit: Card account
        """
        if transaction_type == "purchase":
            lines = [
                {
                    "account_id": merchant_account,
                    "description": f"Card purchase - {transaction_id}",
                    "debit_amount": amount,
                    "credit_amount": 0,
                },
                {
                    "account_id": card_account,
                    "description": f"Card payment",
                    "debit_amount": 0,
                    "credit_amount": amount,
                },
            ]
        elif transaction_type == "refund":
            lines = [
                {
                    "account_id": card_account,
                    "description": f"Card refund",
                    "debit_amount": amount,
                    "credit_amount": 0,
                },
                {
                    "account_id": merchant_account,
                    "description": f"Refund to customer - {transaction_id}",
                    "debit_amount": 0,
                    "credit_amount": amount,
                },
            ]
        else:
            raise ValueError(f"Unsupported transaction type: {transaction_type}")
        
        return await self.create_journal_entry(
            tenant_id=tenant_id,
            user_id=user_id,
            user_role=user_role,
            description=f"Card {transaction_type}: {transaction_id}",
            lines=lines,
            reference=transaction_id,
            metadata={
                "transaction_id": transaction_id,
                "source": "card-service",
                "event_type": transaction_type,
            },
        )
    
    async def record_savings_deposit(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        account_id: str,
        amount: int,
        customer_account: str,
    ) -> Dict[str, Any]:
        """
        Record savings deposit
        
        Debit: Savings account (2200)
        Credit: Customer source account
        """
        lines = [
            {
                "account_id": "2200",  # Savings Deposits
                "description": f"Savings deposit to account {account_id}",
                "debit_amount": amount,
                "credit_amount": 0,
            },
            {
                "account_id": customer_account,
                "description": f"Transfer to savings",
                "debit_amount": 0,
                "credit_amount": amount,
            },
        ]
        
        return await self.create_journal_entry(
            tenant_id=tenant_id,
            user_id=user_id,
            user_role=user_role,
            description=f"Savings deposit: {account_id}",
            lines=lines,
            reference=account_id,
            metadata={
                "account_id": account_id,
                "source": "savings-service",
                "event_type": "deposit",
            },
        )
    
    async def record_savings_withdrawal(
        self,
        tenant_id: str,
        user_id: str,
        user_role: str,
        account_id: str,
        amount: int,
        customer_account: str,
    ) -> Dict[str, Any]:
        """
        Record savings withdrawal
        
        Debit: Customer destination account
        Credit: Savings account (2200)
        """
        lines = [
            {
                "account_id": customer_account,
                "description": f"Savings withdrawal",
                "debit_amount": amount,
                "credit_amount": 0,
            },
            {
                "account_id": "2200",  # Savings Deposits
                "description": f"Withdrawal from account {account_id}",
                "debit_amount": 0,
                "credit_amount": amount,
            },
        ]
        
        return await self.create_journal_entry(
            tenant_id=tenant_id,
            user_id=user_id,
            user_role=user_role,
            description=f"Savings withdrawal: {account_id}",
            lines=lines,
            reference=account_id,
            metadata={
                "account_id": account_id,
                "source": "savings-service",
                "event_type": "withdrawal",
            },
        )

File: services/loan-service/coa_client.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"time"
)

type CoAClient struct {
	baseURL    string
	httpClient *http.Client
}

func NewCoAClient() *CoAClient {
	baseURL := os.Getenv("COA_SERVICE_URL")
	if baseURL == "" {
		gatewayURL := os.Getenv("BACKEND_URL")
		if gatewayURL != "" {
			baseURL = gatewayURL + "/chart-of-accounts"
		} else {
			baseURL = "http://chart-of-accounts-service:8080"
		}
	}
	return &CoAClient{
		baseURL:    baseURL,
		httpClient: &http.Client{Timeout: 60 * time.Second},
	}
}

// PostAsync fires a COA journal entry in a background goroutine.
// Failures are logged but never bubble up to the caller — same pattern as audit.
func (c *CoAClient) PostAsync(tenantID, userID, userRole string, entry CreateJournalEntryRequest) {
	go func() {
		if _, err := c.CreateJournalEntry(tenantID, userID, userRole, entry); err != nil {
			log.Printf("WARN [coa] async journal entry failed (tenant=%s ref=%s): %v", tenantID, entry.Reference, err)
		}
	}()
}

// GetMapping resolves a semantic key (e.g. "loans.interest.sme") to the
// COA account UUID configured by the tenant. Returns "" if not configured.
func (c *CoAClient) GetMapping(tenantID, mappingKey string) string {
	url := fmt.Sprintf("%s/api/v1/mappings/%s", c.baseURL, mappingKey)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return ""
	}
	req.Header.Set("X-Tenant-ID", tenantID)
	req.Header.Set("x-keycloak-id", "system")
	req.Header.Set("X-User-Role", "finance_admin")

	resp, err := c.httpClient.Do(req)
	if err != nil || resp.StatusCode != http.StatusOK {
		return ""
	}
	defer resp.Body.Close()

	var result struct {
		AccountID string `json:"account_id"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return ""
	}
	return result.AccountID
}

// CoAAccount represents an account in the Chart of Accounts
type CoAAccount struct {
	ID              string    `json:"id,omitempty"`
	TenantID        string    `json:"tenant_id"`
	Code            string    `json:"code"`
	Name            string    `json:"name"`
	Type            string    `json:"type"`
	ParentID        string    `json:"parent_id,omitempty"`
	NormalBalance   string    `json:"normal_balance,omitempty"`
	Level           int       `json:"level,omitempty"`
	IsActive        bool      `json:"is_active"`
	IsSystemAccount bool      `json:"is_system_account"`
	Currency        string    `json:"currency"`
	Description     string    `json:"description,omitempty"`
	CreatedAt       time.Time `json:"created_at,omitempty"`
	UpdatedAt       time.Time `json:"updated_at,omitempty"`
}

type JournalLineRequest struct {
	AccountID    string `json:"account_id"`
	Description  string `json:"description,omitempty"`
	DebitAmount  int64  `json:"debit_amount"`
	CreditAmount int64  `json:"credit_amount"`
}

type CreateJournalEntryRequest struct {
	Date        time.Time              `json:"date"`
	Description string                 `json:"description"`
	Reference   string                 `json:"reference,omitempty"`
	Lines       []JournalLineRequest   `json:"lines"`
	PostedBy    string                 `json:"posted_by"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
}

type JournalEntryResponse struct {
	ID          string                 `json:"id"`
	TenantID    string                 `json:"tenant_id"`
	EntryNumber string                 `json:"entry_number"`
	Date        time.Time              `json:"date"`
	Description string                 `json:"description"`
	Status      string                 `json:"status"`
	Lines       []JournalLine          `json:"lines"`
	Metadata    map[string]interface{} `json:"metadata,omitempty"`
	CreatedAt   time.Time              `json:"created_at"`
	UpdatedAt   time.Time              `json:"updated_at"`
}

type JournalLine struct {
	ID           string `json:"id"`
	AccountID    string `json:"account_id"`
	AccountCode  string `json:"account_code"`
	AccountName  string `json:"account_name"`
	Description  string `json:"description"`
	DebitAmount  int64  `json:"debit_amount"`
	CreditAmount int64  `json:"credit_amount"`
}

func (c *CoAClient) CreateJournalEntry(tenantID, userID, userRole string, entry CreateJournalEntryRequest) (*JournalEntryResponse, error) {
	body, err := json.Marshal(entry)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal journal entry: %w", err)
	}

	req, err := http.NewRequest("POST", c.baseURL+"/api/v1/journal-entries", bytes.NewBuffer(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Tenant-ID", tenantID)
	req.Header.Set("x-keycloak-id", userID)
	req.Header.Set("X-User-Role", userRole)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
		bodyBytes, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("failed to create journal entry: %s (status: %d)", string(bodyBytes), resp.StatusCode)
	}

	var result JournalEntryResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}
	return &result, nil
}

// RecordLoanDisbursement posts the journal entry for a loan disbursement.
// Account UUIDs are resolved from the tenant's COA mappings:
//
//	loans.receivable          → asset (Loans Receivable)
//	loans.customer.liability  → liability (Customer Deposits / Nostro)
func (c *CoAClient) RecordLoanDisbursement(tenantID, userID, userRole, loanID, loanType string, amount int64) {
	principalAcct := c.GetMapping(tenantID, "loans.receivable")
	customerAcct := c.GetMapping(tenantID, "loans.customer.liability")

	if principalAcct == "" || customerAcct == "" {
		log.Printf("WARN [coa] disbursement skipped for loan %s — mappings not configured for tenant %s", loanID, tenantID)
		return
	}

	entry := CreateJournalEntryRequest{
		Date:        time.Now(),
		Description: fmt.Sprintf("Loan disbursement - %s", loanID),
		Reference:   loanID,
		PostedBy:    userID,
		Lines: []JournalLineRequest{
			{AccountID: principalAcct, Description: "Loan principal disbursed", DebitAmount: amount},
			{AccountID: customerAcct, Description: "Credit to customer account", CreditAmount: amount},
		},
		Metadata: map[string]interface{}{
			"loan_id": loanID, "loan_type": loanType, "source": "loan-service", "event_type": "disbursement",
		},
	}
	c.PostAsync(tenantID, userID, userRole, entry)
}

// RecordLoanRepayment posts the journal entry for a loan repayment.
// Account UUIDs are resolved from the tenant's COA mappings:
//
//	loans.receivable               → asset (Loans Receivable)
//	loans.interest.<loanType>      → revenue (Interest Income, specific to loan type)
//	loans.customer.liability       → liability (Customer Deposits)
func (c *CoAClient) RecordLoanRepayment(tenantID, userID, userRole, loanID, loanType string, principalAmount, interestAmount int64) {
	principalAcct := c.GetMapping(tenantID, "loans.receivable")
	customerAcct := c.GetMapping(tenantID, "loans.customer.liability")
	interestAcct := c.GetMapping(tenantID, "loans.interest."+loanType)
	if interestAcct == "" {
		// Fall back to generic interest account if loan-type-specific one isn't mapped
		interestAcct = c.GetMapping(tenantID, "loans.interest")
	}

	if principalAcct == "" || customerAcct == "" {
		log.Printf("WARN [coa] repayment skipped for loan %s — mappings not configured for tenant %s", loanID, tenantID)
		return
	}

	totalAmount := principalAmount + interestAmount
	lines := []JournalLineRequest{
		{AccountID: customerAcct, Description: "Loan repayment received", DebitAmount: totalAmount},
		{AccountID: principalAcct, Description: "Principal repayment", CreditAmount: principalAmount},
	}
	if interestAmount > 0 && interestAcct != "" {
		lines = append(lines, JournalLineRequest{
			AccountID: interestAcct, Description: "Interest earned", CreditAmount: interestAmount,
		})
	}

	entry := CreateJournalEntryRequest{
		Date:        time.Now(),
		Description: fmt.Sprintf("Loan repayment - %s", loanID),
		Reference:   loanID,
		PostedBy:    userID,
		Lines:       lines,
		Metadata: map[string]interface{}{
			"loan_id":          loanID,
			"loan_type":        loanType,
			"source":           "loan-service",
			"event_type":       "repayment",
			"principal_amount": principalAmount,
			"interest_amount":  interestAmount,
		},
	}
	c.PostAsync(tenantID, userID, userRole, entry)
}
